src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/NodeFactory.cs

```diff
@@ -101,6 +101,14 @@ private void CreateNodeCaches()
                     {
                         return new GenericDefinitionEETypeNode(this, type);
                     }
+                    else if (type.IsCanonicalDefinitionType(CanonicalFormKind.Any))
+                    {
+                        return new CanonicalDefinitionEETypeNode(this, type);
+                    }
+                    else if (type.IsCanonicalSubtype(CanonicalFormKind.Any))
+                    {
+                        return new CanonicalEETypeNode(this, type);
```

Comment:
Why is constructed and necessary node for canonical types backed by the same implementation? (Necessary nodes shouldn't have a vtable, gcdesc, interface lists - and generally shouldn't ever expect to be newobj'd.)

Canonical EETypes kind of have characteristics of both worlds (shouldn't ever be allocated, but on the other hand, it needs a GCDesc that we can copy).

We should be consistent in how we ask for them - not have code that sometimes asks for it as necessary, sometimes constructed, and gets the same thing anyway.

Should we maybe just put the canonical ones in a separate nodecache that you need to explicitly ask for? (I.e. never return them from Constructed/NecessaryTypeSymbol, but add a CanonicalTypeSymbol method to NodeFactory.)